OutboxSample/Controllers/UserController.cs

```diff
@@ -1,79 +1,90 @@
+        ILogger<UserController> logger) : base(eventMetadataProvider, timeProvider)
     {
         this.userRepository = userRepository;
-        this.outbox = outbox;
         this.logger = logger;
     }
 
     [HttpGet]
-    public IEnumerable<User> Get()
+    public IActionResult Get()
     {
-        IEnumerable<User> users = this.userRepository.GetAll();
-
-        return users;
+        QueryResult<IEnumerable<User>> getAllQueryResult = this.userRepository.GetAll();
 
-        // NOTE: this is just to test IUserRepository.GetAll() from a UnitOfWork
-        // a stupid use-case, but it should work!
-
-        //using (IUnitOfWork work = UnitOfWork.Begin())
-        //{
-        //    var repo = work.GetRepository<IUserRepository>();
+        IActionResult actionResult = getAllQueryResult switch
+        {
+            QueryResult.Success<IEnumerable<User>> success => Ok(success.Data),
 
-        //    IEnumerable<User> users = repo.GetAll();
+            _ => this.UnknownFailure(),
+        };
 
-        //    return users;
-        //}
+        return actionResult;
     }
 
 
     [HttpPost]
-    public IActionResult Post()
+    public IActionResult AddUser([FromServices] IUnitOfWorkFactory unitOfWork)
     {
-        bool saved;
-
-        using (IUnitOfWork work = UnitOfWork.Begin())
+        using (IUnitOfWork work = unitOfWork.Begin())
         {
             var repo = work.GetRepository<IUserRepository>();
 
-            User newUser = new(Guid.NewGuid(), DateTime.Now.ToString());
+            User newUser = new(SequentialUuid.New(), DateTime.Now.ToString());
+
+            QueryResult<int> addQueryResult = repo.Add(newUser);
+
+            if (addQueryResult is QueryResult.Failure)
+            {
+                return addQueryResult switch
+                {
+                    QueryResult<int>.Failure.AlreadyExists failure => base.Conflict($"User with Id=[{newUser.Id}] already exists."),
+
+                    QueryResult<int>.Failure.ConcurrencyConflict failure => base.Conflict(failure.Message),
 
-            repo.Add(newUser);
+                    _ => this.UnknownFailure(),
+                };
+            }
 
             IOutbox outbox = work.GetOutbox();
 
-            outbox.Send(new UserAddedEvent(newUser.Id, newUser.Name));
+            var userAddedEvent = new UserAddedEvent(SequentialUuid.New(), newUser.Id, newUser.Name);
 
-            saved = work.Commit();
-        }
+            // TODO: get agg version from the repo.Add response
+            EventEnvelope envelope = this.WrapEvent(userAddedEvent, rootApplicationAggregateId, 0);
+
+            outbox.Send(envelope);
 
-        return Ok(saved ? "Saved" : "Not saved");
+            bool saved = work.Commit();
+
+            return Ok(saved ? "User added." : "Not added.");
+        }
     }
 
+
     [HttpPost("v2")]
-    public IActionResult PostV2()
+    public IActionResult PostV2([FromServices] IUnitOfWorkFactory unitOfWork)
```

Comment:
TIL about `[FromServices]` 😃 Is this to avoid an allocation of a scoped instance of `IUnitOfWorkFactory` for those controller methods that don't need it?

Reply:
Yup :) Essentially this is to not over-inject your controllers and better 'scope' your dependencies.